include/simulator.h

```diff
@@ -41,18 +42,178 @@ struct CostMetrics {
+private:
+  int num_nodes;
+  int num_gpus_per_node;
+  int num_gpus;
+  float inter_gpu_bandwidth;
+  float inter_node_bandwidth;
+  float gpu_dram_bandwidth;
+  std::map<int, Comp_device*> id_to_gpu;
+  std::map<int, Mem_device*> id_to_gpu_fb_mem;
+  std::map<int, Comm_device*> id_to_gputodram_comm_device;
+  std::map<int, Comm_device*> id_to_dramtogpu_comm_device;
+  std::map<size_t, Comm_device*> ids_to_inter_gpu_comm_device;
+  std::map<size_t, Comm_device*> ids_to_inter_node_comm_device;
+};
+
+class MachineModel_new : public MachineModel {
+public:
+    enum NicDistribution {
+      PER_NODE,
+      PER_SOCKET,
+    };
+    MachineModel_new(std::string file, size_t gpu_fb_mem_capacity);
+    int get_version();
+    Comp_device *get_cpu(int device_id);
+    Comp_device *get_cpu(int socket_id, int local_id);
+    Comp_device *get_gpu(int device_id);
+    Comp_device *get_gpu(int socket_id, int local_id);
+    Mem_device *get_sys_mem(int socket_id);
+    Mem_device *get_z_copy_mem(int socket_id);
+    Mem_device *get_gpu_fb_mem(int device_id);
+    Mem_device *get_gpu_fb_mem(int socket_id, int local_id);
+    Comm_device *get_nvlink(Mem_device *src_mem, Mem_device *tar_mem);
+    int get_num_gpus();
+    float get_intra_node_gpu_bandwidth();
+    float get_inter_node_gpu_bandwidth();
+    std::vector<Comm_device *> get_comm_path(Mem_device *src_mem, Mem_device *tar_mem);
+    std::string to_string();
+private:
+    int num_nodes;
+    int num_sockets_per_node;
+    int num_cpus_per_socket;
+    int num_gpus_per_socket;
+    int num_sockets;
+    int num_cpus;
+    int num_gpus;
+    int num_nvlinks_per_node;
+    float membus_latency;
+    float membus_bandwidth;
+    float upi_latency;
+    float upi_bandwidth;
+    float nic_latency;
+    float nic_bandwidth;
+    NicDistribution nic_distribution;
+    float pci_latency;
+    float pci_bandwidth;
+    float nvlink_latency;
+    float nvlink_bandwidth;
+    size_t gpu_fb_mem_capacity;
+    std::vector<Comm_device::CommDevType> intra_socket_sys_mem_to_sys_mem;
+    std::vector<Comm_device::CommDevType> inter_socket_sys_mem_to_sys_mem;
+    std::vector<Comm_device::CommDevType> inter_node_sys_mem_to_sys_mem;
+    std::vector<Comm_device::CommDevType> intra_socket_sys_mem_to_gpu_fb_mem;
+    std::vector<Comm_device::CommDevType> inter_socket_sys_mem_to_gpu_fb_mem;
+    std::vector<Comm_device::CommDevType> inter_node_sys_mem_to_gpu_fb_mem;
+    std::vector<Comm_device::CommDevType> intra_socket_gpu_fb_mem_to_sys_mem;
+    std::vector<Comm_device::CommDevType> inter_socket_gpu_fb_mem_to_sys_mem;
+    std::vector<Comm_device::CommDevType> inter_node_gpu_fb_mem_to_sys_mem;
+    std::vector<Comm_device::CommDevType> intra_socket_gpu_fb_mem_to_gpu_fb_mem;
+    std::vector<Comm_device::CommDevType> inter_socket_gpu_fb_mem_to_gpu_fb_mem;
+    std::vector<Comm_device::CommDevType> inter_node_gpu_fb_mem_to_gpu_fb_mem;
+    std::vector<std::vector<Comp_device *>> cpus;   // socket_id, local_id
```

Comment:
Would it be reasonable to use a `std::unordered_map` with a struct key that makes these two indices clear, or is there a compelling reason for nested vectors?